File: src/eQuantic.UI.Runtime/src/shared/components/Stepper.ts
import { Box, BoxStyle, BuildContext, CornerRadii, Icon, Pressable, Row, SizeValue, Sizing, StatelessComponent, Text } from "../runtime-exports";

export class Stepper extends StatelessComponent {
    declare value: number;
    declare onChanged: any;
    declare min: number;
    declare max: number;
    declare step: number;
    declare size: string;
    declare disabled: boolean;
    declare suffix: string;
    declare label: string;
    constructor(value?: any, onChanged: any = null, props?: any) {
        super();
        if (value !== undefined) this.value = value;
        if (onChanged !== undefined) this.onChanged = onChanged;
        if (this.min === undefined) this.min = -2147483648;
        if (this.max === undefined) this.max = 2147483647;
        if (this.step === undefined) this.step = 1;
        if (this.size === undefined) this.size = 'medium';
        if (this.suffix === undefined) this.suffix = '';
        if (this.label === undefined) this.label = '';
        this.value = value;this.onChanged = onChanged;
        if (props && typeof props === 'object') Object.assign(this, props);
    }

    build(context: BuildContext) {
        let theme = context.theme;let height = Sizing.height(this.size, context.density);let canDecrement = !this.disabled && this.value - this.step >= this.min;let canIncrement = !this.disabled && this.value + this.step <= this.max;let row = new Row(0, 'start', 'center', false, null, null, { height: SizeValue.fill, cross: 'center' });row.add(Stepper.arm(theme, 'minus', height, canDecrement, () => this.onChanged?.(this.value - this.step), `Decrease ${this.label}`));let reading = new Row(0, 'start', 'center', false, null, null, { height: SizeValue.fill, main: 'center', cross: 'center' });reading.add(new Text(`${this.value}${this.suffix}`, 'label', this.disabled ? theme.textMuted : theme.textPrimary, 1, 'start', false, false, null, { align: 'center', tabular: true, styleOverride: theme.type('label').withSize(Sizing.labelSize(this.size, context.density)) }));row.add(new Box(new BoxStyle({ minWidth: height, height: SizeValue.fill }), reading));row.add(Stepper.arm(theme, 'plus', height, canIncrement, () => this.onChanged?.(this.value + this.step), `Increase ${this.label}`));return new Box(new BoxStyle({ width: SizeValue.hug, height: height, background: this.disabled ? theme.surfaceSubtle : theme.surface, cornerRadius: new CornerRadii(Sizing.radius(this.size)), borderWidth: 1, borderColor: theme.borderStrong, opacity: this.disabled ? theme.disabledOpacity : 1 }), row);
    }

    static arm(theme: any, glyph: string, height: number, enabled: boolean, onPressed: () => void, label: string) {
        let centered = new Row(0, 'start', 'center', false, null, null, { width: SizeValue.fill, height: SizeValue.fill, main: 'center', cross: 'center' });centered.add(new Icon(glyph, 20, enabled ? theme.textPrimary : theme.textMuted));let box = new Box(new BoxStyle({ width: height, height: SizeValue.fill }), centered);return new Pressable(box, enabled ? onPressed : null, { disabled: !enabled, label: label });
    }

}

File: src/eQuantic.UI.Components/Stepper.cs
using eQuantic.UI.Primitives;

namespace eQuantic.UI.Components;

/// <summary>
/// A bounded numeric field driven by two buttons (spec B14), CONTROLLED: the caller owns
/// <see cref="Value"/> and applies <see cref="OnChanged"/>. The frame is a control of
/// <see cref="Size"/> — height, radius and label size come from <see cref="Sizing"/>, so a Stepper
/// lines up with the TextInput beside it. The value reads in TABULAR figures: incrementing must not
/// make the row jiggle. An end of the range disables its button rather than clamping silently, so
/// the limit is visible before it is hit.
/// </summary>
public sealed class Stepper : StatelessComponent
{
    public Stepper(int value, Action<int>? onChanged = null)
    {
        Value = value;
        OnChanged = onChanged;
    }

    public int Value { get; init; }
    public Action<int>? OnChanged { get; init; }

    public int Min { get; init; } = int.MinValue;
    public int Max { get; init; } = int.MaxValue;
    public int Step { get; init; } = 1;
    public SizeVariant Size { get; init; } = SizeVariant.Medium;
    public bool Disabled { get; init; }

    /// <summary>Rendered after the number ("2 GB", "3×") — part of the value, not a separate label.</summary>
    public string Suffix { get; init; } = "";

    /// <summary>Announced by the control; the number alone says nothing about what it counts.</summary>
    public string Label { get; init; } = "";

    public override VisualNode Build(ComponentContext context)
    {
        var theme = context.Theme;
        var height = Sizing.Height(Size);
        var canDecrement = !Disabled && Value - Step >= Min;
        var canIncrement = !Disabled && Value + Step <= Max;

        var row = new Row(gap: 0) { Height = SizeValue.Fill, Cross = CrossAlign.Center };
        row.Add(Arm(theme, Icons.Minus, height, canDecrement,
            canDecrement ? () => OnChanged?.Invoke(Value - Step) : null, $"Decrease {Label}"));

        row.Add(new Flexible(new Text($"{Value}{Suffix}", TypeRole.Label,
            Disabled ? theme.TextMuted : theme.TextPrimary, maxLines: 1)
        {
            Align = TextAlignment.Center,
            Tabular = true,
            StyleOverride = theme.Type(TypeRole.Label) with { Size = Sizing.LabelSize(Size) },
        }, 1));

        row.Add(Arm(theme, Icons.Plus, height, canIncrement,
            canIncrement ? () => OnChanged?.Invoke(Value + Step) : null, $"Increase {Label}"));

        return new Box(new BoxStyle
        {
            Height = height,
            MinWidth = height * 3,
            Background = Disabled ? theme.SurfaceSubtle : theme.Surface,
            CornerRadius = new CornerRadii(Sizing.Radius(Size)),
            BorderWidth = 1,
            BorderColor = theme.BorderStrong,
            Opacity = Disabled ? theme.DisabledOpacity : 1f,
        }, row);
    }

    /// <summary>One end of the control: a square press target the full height of the frame.</summary>
    private static VisualNode Arm(IAppTheme theme, Icons glyph, float height, bool enabled,
        Action? onPressed, string label)
    {
        var centered = new Row(gap: 0) { Main = MainAlign.Center, Cross = CrossAlign.Center };
        centered.Add(new Icon(glyph, IconSize.Dense, enabled ? theme.TextPrimary : theme.TextMuted));

        var box = new Box(new BoxStyle { Width = height, Height = SizeValue.Fill }, centered);

        return new Pressable(box, onPressed) { Disabled = !enabled, Label = label };
    }
}
